torch_ttnn/passes/memory_pass.py

```diff
@@ -0,0 +1,240 @@
+    i = 0
+    overflow_ops = []
+
+    for node in nodes:
+        on_device = False
+        # Has to be a ttnn op
+        if is_tt_compute(node):
+            mm.logs += "---------------------------------------------------\n"
+            print("\n---------------------------------------------------")
+            mm.logs += f"tt compute node numbered: {i}\n"
+            print(f"tt compute node numbered: {i}")
+            i += 1
+            # Exception: Full node has no input
+            # Assumption: Full node outputs a tensor on device
+            if node.target == ttnn.full:
+                on_device = True
+            total_input_size = 0
+            # Input & Output tensor IDs for ttnn op
+            ttnn_ops_tids = []
+            for input_node in node.all_input_nodes:
+                # If input tensor on device or coming as output from another ttnn on device op
+                if op_registry.is_input_tensor_on_device(input_node):
+                    on_device = True
+                    input_shape = op_registry.get_shape(input_node)
+                    input_dtype = op_registry.get_dtype(input_node)
+                    input_size = op_registry.get_size(input_shape, input_dtype)
+                    total_input_size += input_size
+
+                    assert (
+                        input_node in mm.node_to_tid_map
+                    ), "Tensor ID not allocated for one of the inputs!"
+
+                    tid = mm.node_to_tid_map[input_node]
+                    ttnn_ops_tids.append(tid)
+                    mm.tensor_size_of[tid] = input_size
+
+                    start_addr = last_assigned_addr
+                    last_assigned_addr += input_size
+                    end_addr = last_assigned_addr
+                    mm.tid_to_addr_map_in_sram[tid] = (start_addr, end_addr)
+
+                    if tid not in mm.tensors_on_device:
+                        mm.used_sram += input_size
+                        mm.free_sram -= input_size
+                        mm.tensors_on_device.append(tid)
+
+                    if (tid, input_size) not in on_device_meta:
+                        on_device_meta.append((tid, input_size))
+
+            if on_device:
+                output_shape = op_registry.get_shape(node)
+                output_dtype = op_registry.get_dtype(node)
+                output_size = op_registry.get_size(output_shape, output_dtype)
+
+                mm.ops_mem_usage[node.name] = total_input_size + output_size
+
+                assert (
+                    node in mm.node_to_tid_map
+                ), "Tensor ID not allocated for one of the inputs!"
+
+                tid = mm.node_to_tid_map[node]
+                ttnn_ops_tids.append(tid)
+                mm.ttnn_ops_tids[node.name] = ttnn_ops_tids
+                mm.tensor_size_of[tid] = output_size
+
+                start_addr = last_assigned_addr
+                last_assigned_addr += output_size
```

Comment:
This is the exact part which assumes there is no fragmentation, right?

Reply:
I recommend to establish a Memory State and delegate such calculations to it.
In this case 

Something like this
```
memory_state = allocate(memory_state, tensor_id, size)
....
memory_state = deallocate(memory_state, tensor_id)
```

or 
```
memory_state.allocate(tensor_id, size)
memory_state.deallocate(tensor_id)
```

In this case as we evolve the memory_state modeling we can simply substitute the logic inside of that separate structure without touching the high level logic

Reply:
This assumes memory allocation happens in linear fashion meaning if there are deallocations then those deallocated memory regions are not considered for future allocations, hence memory is fragmented.